content/6_Advanced/Count_Min.mdx

```diff
@@ -3,38 +3,214 @@ id: count-min
+- if $a.\texttt{val} < b.\texttt{val}$, then $c = a$
+- if $a.\texttt{val} > b.\texttt{val}$, then $c = b$
+- if $a.\texttt{val} = b.\texttt{val}$, then $c = \{a.\texttt{val}, a.\texttt{cnt} + b.\texttt{cnt}\}$
 
-For each $x$-coordinate, count the number of squares that are covered by
-**none** of the input rectangles.
+### Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+const int MAXN = 200005;
+const int INF = numeric_limits<int>::infinity();
+
+struct node {
+	int val = INF, cnt = 1;
+} T[2 * MAXN];
+
+// combines two segment tree nodes
+node merge(node a, node b) {
+	if (a.val < b.val)
+		return a;
+	if (a.val > b.val)
+		return b;
+	return {a.val, a.cnt + b.cnt};
+}
+
+// updates the ith value to v
+void update(int i, int v) {
+	for (T[i += MAXN] = {v, 1}; i > 1; i >>= 1)
+		T[i >> 1] = merge(T[i], T[i ^ 1]);
+}
+
+// returns the minimum and occurences between indices l and r
+node query(int l, int r) {
+	node res;
+	for (l += MAXN, r += MAXN + 1; l < r; l >>= 1, r >>= 1) {
+		if (l & 1) res = merge(res, T[l++]);
+		if (r & 1) res = merge(res, T[--r]);
+	}
+	return res;
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
+### Solution - Area of Rectangles
+
+<Spoiler title = "Hint 1">
+
+Count the number of squares that are covered by **none** of the input rectangles.
+
+</Spoiler>
+
+<Spoiler title = "Hint 2">
+
+Sort the rectangles by x-coordinate and run a sweep-line.
 
 </Spoiler>
 
 <Spoiler title="Solution">
 
-<IncompleteSection>
+We can use techniques introduced in [Range Queries with Sweep Line](https://usaco.guide/plat/range-sweep?lang=cpp)
+
+We sweep from left to right over the $x$-coordinates, maintaining two events for each rectangle: one for the left boundary and one for the right boundary. Maintain a Lazy Segment Tree over the $y$-coordinates. 
+- When we run into a left boundary of some rectangle with $y$-coordinates $(y_0, y_1)$, increase each index $i \in [y_0, y_1]$ by 1
+- When we run into a right boundary of some rectangle with $y$-coordinates $(y_0, y_1)$, decrease each index $i \in [y_0, y_1]$ by 1
+
+Then, for each $x$, we simply need to count the number of non-zero indices which corresponds to indices that are covered by at least one rectangle. How can we do this?
 
-describe
+Instead of counting the area covered by at least one rectangle, let's count the amount of space covered by no rectangles. We can subtract this amount from the total number of indices to get the value we want. 
 
-</IncompleteSection>
+We can use a Segment Tree that counts the number of occurences of the minimum value. Because the minimum value is at least zero (there can't be a negative number of rectangles at a position) the number of uncovered squares is equal to the number of squares with value 0.
 
 </Spoiler>
 
+### Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int MAXX = 1e6;
+const int MAXN = 2 * MAXX + 5;
```

Comment:
is the +5 really necessary

Reply:
Yeah, there needs to be at least one empty node

Reply:
ok uh then how about this
just put MAXN as 2 million because that's the actual value (presumably)
then when you use it, use MAXN + 1

Reply:
tbh i think this is better 
becuase i have `4 * MAXN` in alot of places and it'll be weird if I put smth like `4 * (MAXN + 1)`

Reply:
why would it be weird?

Reply:
Cuz it's a segment tree

Reply:
it's also common to use macros, you don't see me supporting those